File: calcVAF.R
# read in commented part of vcf (start with at least one #)
openFile = file('/Users/valdezkm/Documents/LP/Val_ClinOmics_Compass_026_T1D_E.strelka.snvs.raw.vcf',open = 'r')
comm = readLines(openFile)
comm = comm[grepl('^#',comm)]
close(openFile)

# read in uncommented vcf table, convert relevant columns to char
dat = read.delim('/Users/valdezkm/Documents/LP/Val_ClinOmics_Compass_026_T1D_E.strelka.snvs.raw.vcf',comment.char = '#', header = FALSE)
dat$V11 = as.character(dat$V11)
dat$V5 = as.character(dat$V5)
dat$V8 = as.character(dat$V8)
dat$VAF = NA

for (i in 1:nrow(dat)) {
  # parse out the number of tumor ACGTs 
  As = as.integer(unlist(strsplit(unlist(strsplit(dat[i,]$V11[1],'\\:'))[5],',')))[1]
  Cs = as.integer(unlist(strsplit(unlist(strsplit(dat[i,]$V11[1],'\\:'))[6],',')))[1]
  Gs = as.integer(unlist(strsplit(unlist(strsplit(dat[i,]$V11[1],'\\:'))[7],',')))[1]
  Ts = as.integer(unlist(strsplit(unlist(strsplit(dat[i,]$V11[1],'\\:'))[8],',')))[1]
  
  # determine depth dependent upon REF variant
  if (dat[i,]$V4=='A') {
    DEPTH = As
  } else if (dat[i,]$V4=='C') {
    DEPTH = Cs
  } else if (dat[i,]$V4=='G') {
    DEPTH = Gs
  } else if (dat[i,]$V4=='T') {
    DEPTH = Ts
  }
  
  # populate VAF column depent upon ALT variant
  dat[i,]$VAF[dat[i,]$V5=='A'] = As / (As + DEPTH)
  dat[i,]$VAF[dat[i,]$V5=='C'] = Cs / (Cs + DEPTH)
  dat[i,]$VAF[dat[i,]$V5=='G'] = Gs / (Gs + DEPTH)
  dat[i,]$VAF[dat[i,]$V5=='T'] = Ts / (Ts + DEPTH)
  
  # recreate info column to include vaf
  dat[i,]$V8 = paste0(dat[i,]$V8,';VAF=',round(dat[i,]$VAF,2))
}
# delete vaf column
dat = subset(dat,select=-VAF)

# write out meta data, append table
write.table(comm,'~/Documents/LP/vaf.vcf',sep='\t',row.names = FALSE,col.names = FALSE,quote = FALSE)
write.table(dat,'~/Documents/LP/vaf.vcf',sep='\t',append = TRUE,row.names = FALSE,quote = FALSE,col.names = FALSE)

File: calcVAF.py
fileN = '/Users/valdezkm/Documents/LP/Val_ClinOmics_Compass_026_T1D_E.strelka.snvs.raw.vcf'

outFile = open('/Users/valdezkm/Documents/LP/gitHub/kv/vaf_py.vcf','w')

with open(fileN) as myfile:
    info = []
    notInfo = []
    chrom = []
    dat = []
    for each in myfile:                                                     # for each line in file
        if each.startswith('##'):                                           # subset commented section
            if each.startswith('##INFO'):                                   # subset info lines, will add vaf later
                info.append(each.strip())
            if not each.startswith('##INFO'):                               # if not info lines, subset
                notInfo.append(each.strip())
        if each.startswith('#CHROM'):                                       # pull out table header
            chrom.append(each.strip())
        if not each.startswith('#'):                                       # subset table and parse ACGTs, use tier 1
            As = float(each.split('\t')[10].split(':')[4].split(',')[0])
            Cs = float(each.split('\t')[10].split(':')[5].split(',')[0])
            Gs = float(each.split('\t')[10].split(':')[6].split(',')[0])
            Ts = float(each.split('\t')[10].split(':')[7].split(',')[0])

            REF = each.split('\t')[3]                                      # Find REF variant, set appropriate value for depth
            if REF == 'A':
                DEPTH = As
            elif REF == 'C':
                DEPTH = Cs
            elif REF == 'G':
                DEPTH = Gs
            elif REF == 'T':
                DEPTH = Ts

            ALT = each.split('\t')[4]                                       # find ALT variant, calc VAF
            if ALT == 'A':
                VAF = round(As / (As + DEPTH), 2)
            elif ALT == 'C':
                VAF = round(Cs / (Cs + DEPTH), 2)
            elif ALT == 'G':
                VAF = round(Gs / (Gs + DEPTH), 2)
            elif ALT == 'T':
                VAF = round(Ts / (Ts + DEPTH), 2)

            sepEach = each.split('\t')
            sepEach[7] = sepEach[7] + ';VAF=' + str(VAF)                    # new info column with vaf
            eachStr = '\t'
            each = eachStr.join(sepEach)                                    # redefine orig each variable
            dat.append(each.strip())

    info.append('##INFO=<ID=VAF,Number=1,Type=Float,Description="Variant Allele Frequency">')   # add info tag vaf to header

    together = notInfo + info + chrom + dat                                                     # put vcf file back together
    for x in together:
        outFile.write(x.strip() + '\n')

